File: solutions/elixir/tournament/1/lib/tournament.ex
defmodule Tournament do
  @doc """
  Given `input` lines representing two teams and whether the first of them won,
  lost, or reached a draw, separated by semicolons, calculate the statistics
  for each team's number of games played, won, drawn, lost, and total points
  for the season, and return a nicely-formatted string table.

  A win earns a team 3 points, a draw earns 1 point, and a loss earns nothing.

  Order the outcome by most total points for the season, and settle ties by
  listing the teams in alphabetical order.
  """
  @header "Team                           | MP |  W |  D |  L |  P"

  @spec tally(input :: list(String.t())) :: String.t()
  def tally([]), do: @header

  def tally(input) do
    stats =
      Enum.reduce(input, %{}, fn line, acc ->
        case parse_line(line) do
          {:ok, {team1, team2, :win}} ->
            acc |> ensure_team(team1) |> ensure_team(team2) |> win(team1, team2)

          {:ok, {team1, team2, :loss}} ->
            acc |> ensure_team(team1) |> ensure_team(team2) |> loss(team1, team2)

          {:ok, {team1, team2, :draw}} ->
            acc |> ensure_team(team1) |> ensure_team(team2) |> draw(team1, team2)

          :error ->
            acc
        end
      end)

    rows =
      stats
      |> Enum.to_list()
      |> Enum.sort_by(fn {name, s} -> {-s.p, name} end)
      |> Enum.map(&format_row/1)

    Enum.join([@header | rows], "\n")
  end

  defp parse_line(line) do
    case String.split(line, ";", parts: 3) do
      [team1, team2, result] ->
        case result do
          "win" -> {:ok, {team1, team2, :win}}
          "loss" -> {:ok, {team1, team2, :loss}}
          "draw" -> {:ok, {team1, team2, :draw}}
          _ -> :error
        end

      _ ->
        :error
    end
  end

  defp ensure_team(acc, team) do
    Map.update(acc, team, %{mp: 0, w: 0, d: 0, l: 0, p: 0}, & &1)
  end

  defp win(acc, w_team, l_team) do
    acc
    |> Map.update!(w_team, fn s -> %{s | mp: s.mp + 1, w: s.w + 1, p: s.p + 3} end)
    |> Map.update!(l_team, fn s -> %{s | mp: s.mp + 1, l: s.l + 1} end)
  end

  defp loss(acc, l_team, w_team) do
    win(acc, w_team, l_team)
  end

  defp draw(acc, team1, team2) do
    acc
    |> Map.update!(team1, fn s -> %{s | mp: s.mp + 1, d: s.d + 1, p: s.p + 1} end)
    |> Map.update!(team2, fn s -> %{s | mp: s.mp + 1, d: s.d + 1, p: s.p + 1} end)
  end

  defp format_row({name, s}) do
    team_width = @header |> String.split("|") |> hd() |> String.length()
    name_field = String.pad_trailing(name, team_width - 1)

    [
      name_field,
      " | ",
      pad2(s.mp),
      " | ",
      pad2(s.w),
      " | ",
      pad2(s.d),
      " | ",
      pad2(s.l),
      " | ",
      pad2(s.p)
    ]
    |> IO.iodata_to_binary()
  end

  defp pad2(int), do: int |> Integer.to_string() |> String.pad_leading(2)
end

File: solutions/ruby/tournament/1/tournament.rb
# Write your code for the 'Tournament' exercise in this file. Make the tests in
# `tournament_test.rb` pass.
#
# To get started with TDD, see the `README.md` file in your
# `ruby/tournament` directory.

class Tournament
  HEADER = 'Team                           | MP |  W |  D |  L |  P'
  ROW_FORMAT = '%-30s | %2d | %2d | %2d | %2d | %2d'

  def self.tally(input)
    team_stats = Hash.new { |hash, team_name| hash[team_name] = default_stats }

    input.to_s.each_line do |line|
      team1, team2, result = parse_line(line)
      next if team1.nil? || team2.nil? || result.nil?

      case result
      when 'win'
        apply_result(team_stats[team1], :win)
        apply_result(team_stats[team2], :loss)
      when 'loss'
        apply_result(team_stats[team1], :loss)
        apply_result(team_stats[team2], :win)
      when 'draw'
        apply_result(team_stats[team1], :draw)
        apply_result(team_stats[team2], :draw)
      end
    end

    lines = [HEADER]
    sorted = team_stats.sort_by { |name, stats| [-stats[:points], name] }
    sorted.each do |name, stats|
      lines << format(ROW_FORMAT, name, stats[:matches], stats[:wins], stats[:draws], stats[:losses], stats[:points])
    end
    lines.join("\n") + "\n"
  end

  class << self
    private

    def default_stats
      { matches: 0, wins: 0, draws: 0, losses: 0, points: 0 }
    end

    def parse_line(line)
      trimmed = line.to_s.strip
      return [nil, nil, nil] if trimmed.empty?

      parts = trimmed.split(';')
      return [nil, nil, nil] unless parts.size == 3

      parts
    end

    def apply_result(stats, result)
      stats[:matches] += 1
      case result
      when :win
        stats[:wins] += 1
        stats[:points] += 3
      when :draw
        stats[:draws] += 1
        stats[:points] += 1
      when :loss
        stats[:losses] += 1
      end
    end
  end
end
